homeassistant/components/google_generative_ai_conversation/__init__.py

```diff
@@ -33,22 +34,42 @@
     CONF_PROMPT,
     DOMAIN,
     RECOMMENDED_CHAT_MODEL,
+    RECOMMENDED_IMAGE_MODEL,
     TIMEOUT_MILLIS,
 )
 
 SERVICE_GENERATE_CONTENT = "generate_content"
+SERVICE_GENERATE_IMAGE = "generate_image"
 CONF_IMAGE_FILENAME = "image_filename"
 CONF_FILENAMES = "filenames"
+CONF_OUTPUT_FILENAME = "output_filename"
 
 CONFIG_SCHEMA = cv.config_entry_only_config_schema(DOMAIN)
 PLATFORMS = (Platform.CONVERSATION,)
 
-type GoogleGenerativeAIConfigEntry = ConfigEntry[genai.Client]
+type GoogleGenerativeAIConfigEntry = ConfigEntry[Client]
 
 
 async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
     """Set up Google Generative AI Conversation."""
 
+    async def upload_attachments(
+        client: Client, filenames: Iterable[str]
+    ) -> list[types.File]:
+        """Upload attachments and return a list of parts."""
+        parts = []
+        for filename in filenames:
+            if not hass.config.is_allowed_path(filename):
+                raise HomeAssistantError(
+                    f"Cannot read `{filename}`, no access to path; "
+                    "`allowlist_external_dirs` may need to be adjusted in "
+                    "`configuration.yaml`"
+                )
+            if not Path(filename).exists():
+                raise HomeAssistantError(f"`{filename}` does not exist")
+            parts.append(await client.aio.files.upload(file=filename))
```

Comment:
Can you introduce back the change to open the file before sending it to the SDK?

Reply:
Why? 

This should be better. 
For example, the SDK might stream it (now or in the future; I didn't check)

Reply:
@IvanLH can you revert your change in a separate PR? I didn't realize that now we have to fit the whole file in memory. It's better to stream it if possible. Especially since your previous PR didn't actually fix the issue it was trying to fix

Reply:
This PR reverts that anyway...

Reply:
From https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr : Make your PRs as small as possible

Reply:
Where are we loading the whole file into memory? We are doing file open, which gets us a handle on the file, the SDK then does seek, which according to my googling doesn't load the file either. Afterwards the SDK does indeed do file read in 8MB chunks. Hoewever this behavior is the same when giving it a file name. The open() is just one layer down.

As for the issue, it is true it seems that it was not fixed by the code. But the logic of the fix is sound, os.path can be wrong. Which would lead to a wrong size being computed. I do agree this is a tight corner since it would require the Filesystem to have a value different from the actual disk size.

My plan was to keep it in place until the SDK fixes it internally.

Reply:
I assumed we were loading the whole file in memory based on SLaks' comment about streaming. Keeping the current checked in behavior is fine. @SLals can you update your code?